[SPARK-9803] [SPARKR] Add subset and transform + tests

felixcheung opened the pull request:
Add subset and transform
Also reorganize `[` & `[[` to subset instead of select

Note: for transform, transform is very similar to mutate. Spark doesn't seem to replace existing column with the name in mutate (ie. `mutate(df, age = df$age + 2)` - returned DataFrame has 2 columns with the same name 'age'), so therefore not doing that for now in transform.
Though it is clearly stated it should replace column with matching name (should I open a JIRA for mutate/transform?)

felixcheung committed:
Add subset and transform + tests
Move [ & [[ to subset instead of select

AmplabJenkins commented:
Can one of the admins verify this patch?

felixcheung commented:
@shivaram

shivaram commented:
Jenkins, ok to test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #41747 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41747/consoleFull) for   PR 8503 at commit [`17472b9`](https://github.com/apache/spark/commit/17472b9115a67fedf04237c6a4c66575ce4c39d3).

shivaram commented:
cc @sun-rui , @yu-iskw (especially for the rdname changes)

SparkQA commented:
[Test build #41747 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41747/console) for   PR 8503 at commit [`17472b9`](https://github.com/apache/spark/commit/17472b9115a67fedf04237c6a4c66575ce4c39d3).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41747/
Test FAILed.

felixcheung commented:
Hmm, I'm guessing the generic is breaking signature matching somehow, with
`result <- transform(Ozone, logOzone = log(Ozone))`
I was running run-tests.sh locally and it passes. Though it looks like this test requires plyr - I'll re-run it then - can we make this a requirement in run-tests.sh?

shivaram commented:
I don't think we should add more dependencies for tests and I don't see why we need plyr. The `transform` function in R is defined at https://stat.ethz.ch/R-manual/R-devel/library/base/html/transform.html and one thing is that the generic / method we define needs to exactly match the definition there (including the names of variables etc.)

felixcheung commented:
This test doesn't run unless plyr is installed. Not sure why.
I think the mutate signature doesn't match either. I will update them.





On Fri, Aug 28, 2015 at 12:53 PM -0700, "Shivaram Venkataraman" <notifications@github.com> wrote:
I don't think we should add more dependencies for tests and I don't see why we need plyr. The `transform` function in R is defined at https://stat.ethz.ch/R-manual/R-devel/library/base/html/transform.html and one thing is that the generic / method we define needs to exactly match the definition there (including the names of variables etc.)

---
Reply to this email directly or view it on GitHub:
https://github.com/apache/spark/pull/8503#issuecomment-135872342

shivaram commented:
mutate is only defined in `plyr` and not in base R, so we have less of a need to match its signature (though I guess it will be good to do that)

felixcheung committed:
fix method signature for transform and mutate to fix mapping; add test
not changing the test_includePackage test but it currently check for plyr package to run

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #41767 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41767/consoleFull) for   PR 8503 at commit [`90916d1`](https://github.com/apache/spark/commit/90916d12f155d74d6a75710a6c1a4c1fa3acada1).

shivaram commented on R/pkg/inst/tests/test_sparkSQL.R:
```diff
@@ -1042,6 +1046,19 @@ test_that("mutate(), rename() and names()", {
+  expect_equal(first(filter(transformedDF, transformedDF$name == "Andy"))$newAge, -30)
+
+  # ensure the proper signature is used - otherwise this will fail to run
+  attach(airquality)
```
Hmm so this is testing whether `transform` on local data frames continue to work ? its an interesting idea -- could you update the comment on top to say `test if transform on local data frames works. ensures that proper signature ...`.

shivaram commented:
@felixcheung for the column being replaced part, I'd say we can open another jira / discuss it for later. I prefer the existing behavior of appending columns and users can drop columns (either with a select or by setting it to `NULL` in SparkR)

cc @falaki

felixcheung committed:
update comment in test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #41768 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41768/consoleFull) for   PR 8503 at commit [`3c74245`](https://github.com/apache/spark/commit/3c74245e6426fffc8a57e89dd62cc2d46784a715).

SparkQA commented:
[Test build #41767 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41767/console) for   PR 8503 at commit [`90916d1`](https://github.com/apache/spark/commit/90916d12f155d74d6a75710a6c1a4c1fa3acada1).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41767/
Test PASSed.

yu-iskw commented on R/pkg/R/DataFrame.R:
```diff
@@ -1034,6 +1068,10 @@ setMethod("[", signature(x = "DataFrame", i = "Column"),
 #' @export
 #' @rdname select
+#' @name select
+#' \seealso{
```
roxygen2 has `@seealso` tag.

felixcheung replied:
should we use `@seealso` or `@family`?

yu-iskw replied:
It would be better to use `@family`.

felixcheung replied:
added @family.

yu-iskw commented on R/pkg/R/generics.R:
```diff
@@ -467,7 +467,7 @@ setGeneric("merge")
 #' @rdname withColumn
 #' @export
-setGeneric("mutate", function(x, ...) {standardGeneric("mutate") })
+setGeneric("mutate", function(.data, ...) {standardGeneric("mutate") })
```
I'm not sure we should use `.` prefix for a parameter.

felixcheung replied:
that's the signature from the original definition. please see discussion above.

yu-iskw replied:
I see. Thanks.

yu-iskw commented on R/pkg/R/generics.R:
```diff
@@ -507,6 +507,10 @@ setGeneric("saveAsTable", function(df, tableName, source, mode, ...) {
 
+#' @rdname withColumn
+#' @export
+setGeneric("transform", function(`_data`, ...) {standardGeneric("transform") })
```
I don't think we should use a quoted parameter. How about `data`?

felixcheung replied:
that's the signature from the original definition. please see discussion above.
local data.frame or vector will fail to run with transform if this doesn't match exactly.

yu-iskw replied:
Oh, I see.

SparkQA commented:
[Test build #41768 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41768/console) for   PR 8503 at commit [`3c74245`](https://github.com/apache/spark/commit/3c74245e6426fffc8a57e89dd62cc2d46784a715).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41768/
Test PASSed.

felixcheung committed:
update documentation reference for @family

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #41775 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41775/consoleFull) for   PR 8503 at commit [`7cf9259`](https://github.com/apache/spark/commit/7cf925984d220dd8f06300dc5012ed8fe43bc5ee).

felixcheung commented:
Opened an issue to track to replace behavior
https://issues.apache.org/jira/browse/SPARK-10346

SparkQA commented:
[Test build #41775 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41775/console) for   PR 8503 at commit [`7cf9259`](https://github.com/apache/spark/commit/7cf925984d220dd8f06300dc5012ed8fe43bc5ee).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41775/
Test PASSed.

shivaram commented:
LGTM. @yu-iskw Any other comments ?

yu-iskw commented:
LGTM

shivaram commented:
Thanks @felixcheung @yu-iskw -- Merging this to master and branch-1.5
